File: scripts/topology_proxy.py
# SPDX-License-Identifier: Apache-2.0
import os
import sys
import threading
from pathlib import Path
from psp_cdl_mcp_server import McpServer
from psp_cdl_mcp_server.stdio import serve_stdio
from psp_cdl_mcpproxy import McpDispatchGate
from psp_cdl_mcpproxy.mcp import StdioMcpClient, create_mcp_proxy_service
from dispatch_fixtures import Fixture
from topology_common import case_by_id, approval, environment, EventLog, close_approval
from fixture_isolation import isolate

case_id, executable, script, server_spy, proxy_spy = sys.argv[1:]
case = case_by_id(case_id)
log = EventLog(proxy_spy,case_id)
peer = f = None
timer = threading.Timer(30,lambda:os._exit(2))
timer.daemon = True
timer.start()
closed = False
try:
    f = Fixture({'agents':'mcp://reference/allowed','temporaryRoot':str(Path(proxy_spy).parent)})
    peer = StdioMcpClient.connect({'executable':executable,'args':[script,'C',case_id,server_spy],'env':environment(),
        'serverInfo':{'name':'psp-cdl-reference','version':'0.1.0'},'timeoutMs':4000})
    registrations = peer.registrations('reference',[approval(name,['used-for-model-training'] if case['input']['perturbation'] == 'proxy-policy-denial' else []) for name in ('allowed','other')],f.now)
    gate = McpDispatchGate(f.store,f,'registry-1',registrations)
    service = create_mcp_proxy_service(gate,f.session['sessionId'],lambda:f.options)
    control = peer.registrations('fixture-control',[close_approval()],f.now)[0]
    def shutdown():
        global closed, peer, f
        if not closed:
            if not Path(server_spy+'.closed').exists(): control['invoke']({},f.options)
            closed = True
            peer.close()
            peer = None
            f.close()
            f = None
            log.close()
            threading.Timer(1,lambda:os._exit(0)).start()

    class Wrapped:
        def authenticate(self, token): return service.authenticate(token)
        def discover(self, *_): return [{k:r[k] for k in ('name','inputSchema','outputSchema')} for r in registrations] + [close_approval()]
        def call_tool(self, name, args, token, principal):
            global closed, peer, f
            if name == 'fixture_close' and args == {}:
                shutdown()
                return {'data':{'closed':True}}
            log.write('proxy-dispatch',args.get('recordId'))
            try:
                result = service.call_tool('reference.'+name,args,token,principal)
                log.write('proxy-release',args['recordId'],'ALLOW')
                return result
            except Exception as exc:
                log.write('proxy-denied',args.get('recordId'),getattr(exc,'code','INTERNAL_ERROR'))
                shutdown()
                raise
    isolate()
    log.write('isolation-probes-blocked')
    serve_stdio(McpServer(Wrapped(),lambda:'test-owner'))
finally:
    timer.cancel()
    try:
        if peer: peer.close()
    finally:
        if f: f.close()
        if not closed: log.close()

File: scripts/topology-proxy.mjs
// SPDX-License-Identifier: Apache-2.0
import {McpServer} from '@psp-cdl/mcp-server';
import {serveStdio} from '@psp-cdl/mcp-server/stdio';
import {McpDispatchGate} from '@psp-cdl/mcpproxy';
import {StdioMcpClient,createMcpProxyService} from '@psp-cdl/mcpproxy/mcp';
import {fixture} from './dispatch-fixtures.mjs';
import {caseById,approval,environment,eventLog,closeApproval} from './topology-common.mjs';
import {isolate} from './fixture-isolation.mjs';
import {dirname} from 'node:path';
import {existsSync} from 'node:fs';
const [id,executable,script,serverSpy,proxySpy]=process.argv.slice(2),c=caseById(id);
const log=eventLog(proxySpy,id);let peer,f;
const timer=setTimeout(()=>process.exit(2),30000);
let closed=false;
try {
  f=await fixture({agents:'mcp://reference/allowed',temporaryRoot:dirname(proxySpy)});
  peer=await StdioMcpClient.connect({executable,args:[script,'C',id,serverSpy],env:environment(),serverInfo:{name:'psp-cdl-reference',version:'0.1.0'},timeoutMs:4000});
  const registrations=peer.registrations('reference',['allowed','other'].map(name=>approval(name,c.input.perturbation==='proxy-policy-denial'?['used-for-model-training']:[])),f.host.now);
  const gate=new McpDispatchGate(f.store,f.host,'registry-1',registrations);
  const service=createMcpProxyService(gate,f.session.sessionId,()=>f.options);
  const control=peer.registrations('fixture-control',[closeApproval()],f.host.now)[0];
  const shutdown=async()=>{
    if(!closed){
      if(!existsSync(serverSpy+'.closed'))await control.invoke({},f.options);
      closed=true;await peer.close();peer=null;f.close();f=null;log.close();setTimeout(()=>process.exit(0),1000);
    }
  };
  // Fixed wire names bridge the local namespaced gate; discovery is not authorization.
  const wrapped={authenticate:t=>service.authenticate(t),
    discover:()=>[...registrations.map(r=>({name:r.name,inputSchema:r.inputSchema,outputSchema:r.outputSchema})),closeApproval()],
    callTool:async(name,args,token,p)=>{
      if(name==='fixture_close'&&Object.keys(args).length===0) {
        await shutdown();
        return {data:{closed:true}};
      }
      log.write('proxy-dispatch',args.recordId??null);
      try {
        const result=await service.callTool('reference.'+name,args,token,p);
        log.write('proxy-release',args.recordId,'ALLOW');return result;
      }catch(error){log.write('proxy-denied',args.recordId??null,error.code??'INTERNAL_ERROR');await shutdown();throw error;}
    }};
  isolate();log.write('isolation-probes-blocked');
  await serveStdio(new McpServer(wrapped,()=> 'test-owner'));
}finally {clearTimeout(timer);try {await peer?.close();}finally{f?.close();if(!closed)log.close();}}
